pkg/controller/dnszone/dnszone_controller.go

```diff
@@ -522,3 +525,37 @@ func IsErrorUpdateEvent(evt event.UpdateEvent) bool {
 
 	return false
 }
+
+// IsClusterDeploymentDNSErrorUpdateEvent returns true when the update event in ClusterDeployment
+// is from DNS error state.
+func IsClusterDeploymentDNSErrorUpdateEvent(evt event.UpdateEvent) bool {
+	new, ok := evt.ObjectNew.(*hivev1.ClusterDeployment)
+	if !ok {
+		return false
+	}
+	if len(new.Status.Conditions) == 0 {
+		return false
+	}
+
+	old, ok := evt.ObjectOld.(*hivev1.ClusterDeployment)
+	if !ok {
+		return false
+	}
+
+	cn := controllerutils.FindClusterDeploymentCondition(new.Status.Conditions, hivev1.DNSNotReadyCondition)
+	if cn != nil && cn.Status == corev1.ConditionTrue {
+		co := controllerutils.FindClusterDeploymentCondition(old.Status.Conditions, hivev1.DNSNotReadyCondition)
+		if co == nil {
+			return true // newly added failure condition
+		}
```

Comment:
is this check required since we are initializing the ClusterDeployment with all the conditions now?

Reply:
update events include the old and new object, and therefore there can be a case where old has no condition set.